modules/age.nix

```diff
@@ -194,57 +213,95 @@ in {
-      text = newGeneration;
-      deps = [
-        "specialfs"
+  config = mkIf (cfg.secrets != {}) (mkMerge [
+    {
+      assertions = [
+        {
+          assertion = cfg.identityPaths != [];
+          message = "age.identityPaths must be set.";
+        }
       ];
-    };
+    }
 
-    system.activationScripts.agenixInstall = {
-      text = installSecrets;
-      deps = [
-        "agenixNewGeneration"
-        "specialfs"
-      ];
-    };
+    (optionalAttrs (!isDarwin) {
+      # Create a new directory full of secrets for symlinking (this helps
+      # ensure removed secrets are actually removed, or at least become
+      # invalid symlinks).
+      system.activationScripts.agenixNewGeneration = {
+        text = newGeneration;
+        deps = [
+          "specialfs"
+        ];
+      };
 
-    # So user passwords can be encrypted.
-    system.activationScripts.users.deps = ["agenixInstall"];
+      system.activationScripts.agenixInstall = {
+        text = installSecrets;
+        deps = [
+          "agenixNewGeneration"
+          "specialfs"
+        ];
+      };
 
-    # Change ownership and group after users and groups are made.
-    system.activationScripts.agenixChown = {
-      text = chownSecrets;
-      deps = [
-        "users"
-        "groups"
-      ];
-    };
+      # So user passwords can be encrypted.
+      system.activationScripts.users.deps = ["agenixInstall"];
 
-    # So other activation scripts can depend on agenix being done.
-    system.activationScripts.agenix = {
-      text = "";
-      deps = ["agenixChown"];
-    };
-  };
+      # Change ownership and group after users and groups are made.
+      system.activationScripts.agenixChown = {
+        text = chownSecrets;
+        deps = [
+          "users"
+          "groups"
+        ];
+      };
+
+      # So other activation scripts can depend on agenix being done.
+      system.activationScripts.agenix = {
+        text = "";
+        deps = ["agenixChown"];
+      };
+    })
+    (optionalAttrs isDarwin {
+      system.activationScripts = {
+        # Secrets with root owner and group can be installed before users
+        # exist. This allows user password files to be encrypted.
+        preActivation.text = builtins.concatStringsSep "\n" [
+          newGeneration
+          installSecrets
+        ];
+
+        # Other secrets need to wait for users and groups to exist.
+        users.text = lib.mkAfter ''
+          ${chownSecrets}
+        '';
+      };
+
+      launchd.daemons.activate-agenix = {
+        script = ''
+          set -e
+          set -o pipefail
+          export PATH="${pkgs.gnugrep}/bin:${pkgs.coreutils}/bin:@out@/sw/bin:/usr/bin:/bin:/usr/sbin:/sbin"
+          ${newGeneration}
+          ${installSecrets}
+          ${chownSecrets}
+          exit 0
+        '';
+        serviceConfig.RunAtLoad = true;
+        serviceConfig.KeepAlive.SuccessfulExit = false;
+      };
```

Comment:
Why do you need the activation scripts and the daemon? Is it because nix-darwin does not run activation scripts on boot?

Reply:
Good Q, I'll see if I can sort that out this morning.

Reply:
> nix-darwin does not run activation scripts on boot

Configurable, it does by default: https://daiderd.com/nix-darwin/manual/index.html#opt-services.activate-system.enable

So perhaps not needed? Will experiment.

Reply:
AFAIU the problem with nix-darwin on-boot activation is that it doesn't allow any customizations, so it isn't possible (or at least not very straightforward) to activate agenix together with the rest of nix-darwin.

Reply:
@ryantm -- @rtimush is right, after a little testing it looks like the situation is:

- the `activationScripts` are run on `darwin-rebuild switch` (confirmed)
- the launchd script is run on `darwin-rebuild switch` (confirmed)
- the `activationScripts` look like they would *not* be run on reboot (untested)
- the launchd script is run on reboot

So we could remove the `activationScripts` stuff I guess? Though it seems like it's the "right" place to put it...

Reply:
If I remember correctly, I actually occasionally had problems when agenix was both integrated in `activationScripts` and as a launchd script, because they try to execute concurrently. So, I'd say from a practical standpoint keeping only a launchd script seems to be the best option.

Reply:
Updated above and confirmed which are / are not run after reboot.

Unfortunately, while the launchd script *was* run, it **did not** set up my secrets. Removing the activationScripts should help in debugging this though!